packages/interactor/src/when.ts

```diff
@@ -0,0 +1,51 @@
+interface IOptions {
+  timeout?: number;
+  message?: string;
+}
+interface When {
+  <Value>(effect: () => Promise<Value> | Value, options?: IOptions): Promise<Value>;
+  timeout: number;
+}
+
+function elapsedSince(start: number) {
+  return Date.now() - start;
+}
+
+export const when: When = async function when<Value>(
+  proc: () => Promise<Value> | Value,
+  { timeout = when.timeout }: IOptions = { timeout: when.timeout }
+): Promise<Value> {
+  let start = Date.now();
+  let success = false;
+  let result: Value;
+  let error: Error;
+
+  async function attempt() {
+    if (elapsedSince(start) >= timeout) {
+      return;
+    }
+
+    try {
+      result = await proc();
+      success = true;
+    } catch (e) {
+      error = e;
+      await new Promise(resolve => {
+        setTimeout(async () => {
+          await attempt();
+          resolve();
+        }, 10);
+      });
```

Comment:
I'm wondering if an iterative solution wouldn't be easier to understand, rather than using recursion? I think if we have some helper function `sleep(x)`, which just returns a promise which resolves in x milliseconds, then we can write it like this:

``` javascript
while(true) {
  await sleep(10);
  if (elapsedSince(start) >= timeout) { break }
  // ...
}
```

Reply:
Awhile back, I made a prototype based on async iterators that provided a pretty elegant approach https://github.com/cowboyd/convergence/blob/master/src/converge.js